### ISSUE TITLE
Updates to Working Mode/Test Review Process

### ISSUE THREAD

**Issue by jscholes** [w3c/aria-at#375]
Members of the PAC team are preparing and developing test plans for various APG patterns, with other members of the community group also having declared an interest in potentially submitting test data.  At the time of writing, there are seven open pull requests representing test plans which are ready for community review, with the oldest (https://github.com/w3c/aria-at/pull/330) open since November 10 2020.

As each test plan is developed on its own branch, there exists a distinct possibility of those branches becoming significantly out of sync with master/main (and each other) as they sit in the "internal review" column on the project board.  In turn, this increases the likelihood of the need to address merge conflicts down the line, or update the test plans to reflect process or technical changes introduced since their initial creation.  It also introduces the need for duplication, e.g. to update shared resources like `keys.mjs` multiple times to add the same commands across multiple branches.

With the above in mind, it is believed that updates to the [Working Mode document](https://github.com/w3c/aria-at/wiki/Working-Mode) are needed to set additional expectations around how and when test reviews will take place.  The Tests -> Process section of that document currently states:

> 6. (Community Group) Internal review of the proposed assertions, assertion priorities, and expected user settings.

This matches the "Internal review (step 6)" column of the [Test Plan Workflow project board](https://github.com/w3c/aria-at/projects/1), but is light on detail.  The current route forward, as proposed by @mcking65 and briefly discussed during the December 17 2020 CG call, is as follows:

1. In-place review is carried out before merging to master/main as soon as possible.  This could be carried out by one community group member.
2. After a branch has been merged into master/main, other members of the community group are encouraged to run and review the tests within the ARIA-AT app directly, as it provides a facility for easily reporting any issues relating to a specific test.  The main GitHub issue corresponding to the pattern under test will also be available for ongoing discussion.
3. If needed, follow-up pull requests can be submitted to make any necessary revisions, which should also be merged as soon as is feasible, and community group members can then rerun the updated tests.  This approach is viable because the ARIA-AT app retains result data for tests which haven't changed and allows selective deletion of data which is no longer relevant.
4. Once the community group is satisfied that sufficient review has been carried out without conflicting results, the test plan can be published in draft form to facilitate feedback from AT developers.

The above does raise some outstanding questions:

1. Should initial CG review only be carried out by someone without a close association with the test developer(s)?  E.g. a member of a particular team or professional organisation shouldn't review tests submitted by other members of the same team or org.  This may need to be formally defined.
2. For point `#4` above, what represents "sufficient review"?  Should this be based on a timeline, a particular number of CG members having executed the tests, something else?

Appreciate any thoughts and further discussion about how we can keep the workflow moving along.

**Comment by mcking65** [comment]
> 1. In-place review is carried out before merging to master/main as soon as possible.  This could be carried out by one community group member.

Suggest: New and changed tests may be merged to main branch at editor's discretion after a reasonable quality check is performed by a qualified independent reviewer.

> 2. After a branch has been merged into master/main, other members of the community group are encouraged to run and review the tests within the ARIA-AT app directly, as it provides a facility for easily reporting any issues relating to a specific test.  The main GitHub issue corresponding to the pattern under test will also be available for ongoing discussion.
> 3. If needed, follow-up pull requests can be submitted to make any necessary revisions, which should also be merged as soon as is feasible, and community group members can then rerun the updated tests.  This approach is viable because the ARIA-AT app retains result data for tests which haven't changed and allows selective deletion of data which is no longer relevant.
> 4. Once the community group is satisfied that sufficient review has been carried out without conflicting results, the test plan can be published in draft form to facilitate feedback from AT developers.

All of 2, 3, and 4 can become a single step in the working mode process:

Community group members run tests to achieve two goals: 1) Ensure the tests are sufficiently accurate and complete to publish as draft expectations that are ready for stakeholder review; 2) Generate draft test data that is ready for stakeholder review. Editors may publish draft tests and test data at their discretion after at least two community group members have generated identical results. Editors are responsible for ensuring the quality and completeness of reviews. Reviewers are responsible for raising issues against tests where they do not meet community group quality guidelines.

> 1. Should initial CG review only be carried out by someone without a close association with the test developer(s)?  E.g. a member of a particular team or professional organisation shouldn't review tests submitted by other members of the same team or org.  This may need to be formally defined.
> 2. For point `#4` above, what represents "sufficient review"?  Should this be based on a timeline, a particular number of CG members having executed the tests, something else?

I don't think we have to answer all this in the working mode itself. We may want to separately establish quality guidelines and put the onus on editors.

**Comment by jscholes** [comment]
The proposed wording will be added to an extract of the Working Mode page and posted here, for people to read and comment on in context.

**Comment by jscholes** [comment]
I've edited steps 6 and 7 of the Working Mode page in line with Matt's suggestions.  Please review below and provide feedback.

Steps 5 through 8 now read:

5. (Test developer) [Write a test plan](https://github.com/w3c/aria-at/wiki/How-to-contribute-tests) that proposes assertions, assertion priorities, and expected user settings, that:
    *   Are aligned with the intent of the ARIA specification and APG documentation.
    *   Are aligned with what ATs already do or what the test developer believes the AT developers are willing to implement.
6. (Community Group) New and changed tests may be merged to main branch at editor's discretion after a reasonable quality check is performed by a qualified independent reviewer.
7. (Community Group) Community group members run tests to achieve two goals:
    1. Ensure the tests are sufficiently accurate and complete to publish as draft expectations that are ready for stakeholder review;
    2. Generate draft test data that is ready for stakeholder review.
    
    Editors may publish draft tests and test data at their discretion after at least two community group members have generated identical results. Editors are responsible for ensuring the quality and completeness of reviews. Reviewers are responsible for raising issues against tests where they do not meet community group quality guidelines.
8. (AT developer, optionally Web developer) Review the proposed assertions, assertion priorities, and expected user settings (within 3 weeks prior to next full test cycle)

#### Full Process Section

1. (Test developer) Select a GitHub issue specifying an ARIA design pattern implementation for which a test plan is needed from the “[Test Plan Workflow](https://github.com/w3c/aria-at/projects/1)” GitHub project’s backlog. All documentation of the test plan development will be recorded in the GitHub issue.
2. (Test developer) Read the relevant parts of the [ARIA specification](https://w3c.github.io/aria/) and the [APG](https://w3c.github.io/aria-practices/) to understand the intent of the accessibility semantics and features present in the implementation that will be tested. Document understanding and provide links in the GitHub issue.
3. (Test developer) If there are native OS components with similar accessibility semantics, study how AT behaves when using those components. Document relevant findings in the GitHub issue.
4. (Test developer) Research any special AT configuration requirements. If some relevant accessibility semantics are not supported using the default AT configuration, explore and document the AT configuration changes needed to convey the semantics. Document relevant findings in the GitHub issue.
5. (Test developer) [Write a test plan](https://github.com/w3c/aria-at/wiki/How-to-contribute-tests) that proposes assertions, assertion priorities, and expected user settings, that:
    *   Are aligned with the intent of the ARIA specification and APG documentation.
    *   Are aligned with what ATs already do or what the test developer believes the AT developers are willing to implement.
6. (Community Group) New and changed tests may be merged to main branch at editor's discretion after a reasonable quality check is performed by a qualified independent reviewer.
7. (Community Group) Community group members run tests to achieve two goals:
    1. Ensure the tests are sufficiently accurate and complete to publish as draft expectations that are ready for stakeholder review;
    2. Generate draft test data that is ready for stakeholder review.
    
    Editors may publish draft tests and test data at their discretion after at least two community group members have generated identical results. Editors are responsible for ensuring the quality and completeness of reviews. Reviewers are responsible for raising issues against tests where they do not meet community group quality guidelines.
8. (AT developer, optionally Web developer) Review the proposed assertions, assertion priorities, and expected user settings (within 3 weeks prior to next full test cycle)
9. (Test developer) Address any review feedback.
10. (Test admin) Initiate a full test cycle or a partial test cycle. Assign testers to different test plan runs.
11. (Tester) Run the test as part of a test cycle. Each test is run by two testers.
12. (Tester) Report raw test results.
13. (Tester, Test admin) Resolve any differences between raw results from different testers.
14. (Test admin) Report draft test results.
15. (AT developer, optionally Web developer) Review the draft test results. (Timebox: TODO)
    *   Provide feedback if any in the relevant GitHub issue.
    *   (Test admin) Mark test results as reviewed if they are OK.
    *   AT developers may change their implementation and ask the Community Group to re-run the test in a new build/version.
16. (Community Group) Address any feedback on the draft test results. This can include changing test plans, tests, and re-running tests.
17. (Community Group) Report reviewed test results.
18. (Community Group) If there are issues in the ARIA specification, the APG document, browser implementations, or OS-level accessibility API implementations, then file an issue for the ARIA Working Group or APG Task Force as appropriate.
19. (AT developer) If there is a bug in AT, file the bug for the AT.
20. (AT developer) Fix the bug, or provide (late) feedback on the test.

**Comment by jscholes** [comment]
Leaving this open potentially until we talk more about #300 (status of a test) and have more resolutions from that that we can work into the Working Mode doc.

**Comment by mcking65** [comment]
I have been working on a [new version of the working mode](https://github.com/w3c/aria-at/wiki/Working-Mode-2021-Update-Draft) to resolve concerns raised in this issue and others. In addition to lowering the bar for merging a test plan into the main branch so it can be exercised by the community group while still a draft, these changes also:
* more clearly define interaction with AT developers.
* Timebox the AT developer review process.
* Describe how automation fits into the process.
* Address how product versions are managed.


I moved definition of terms to a new glossary page, and added several terms.

I broke the process into 3 primary phases in the life cycle of a test plan (draft, candidate, and recommended). As I write this out, I now realize that we need a phase that brings a test plan to end of life in the event it becomes obsolete.

This is still not a complete draft, but it is definitely ready for review and discussion.

**Comment by zcorpan** [comment]
Thanks, @mcking65 

A few questions and comments.

> **Recommended test phase**: Test plans are executed to generate reports. Anyone may raise issues if they believe they have discovered an error in a report. The community group will triage report issues within 30 days, and if it concurs with the issue, will retract the data and resolve the issue, which could return the test plan to the candidate phase. Retractions are publicly logged.

Why retract data? It seems better to me to leave data as-is, but possibly annotate it with links to known issues.

In the Draft Test Phase, step 9:
> Remove assertions effected by the blocked semantic.
> Catalog each assertion that was removed in the test plan development issue and reference the appropriate blocked semantic issue.

Why remove? It seems better to leave the assertions in, so that when the issue is fixed, we can see if the assertion gets the expected result. Without the assertion it's not possible to track progress of the issue in terms of test results.

In the Recommended test phase, step 6:
> (Test developer) Configure automated testing for the test plan.
It would be nice to automate tests earlier, why not in the Draft Test Phase?

More high-level, I'm a bit concerned over the phases and the many steps. In my opinion, this was a problem with the previous working mode, too, but the new version doesn't fix it, instead adds to it.

If we want AT developers to contribute not only by reviewing tests, but also by writing new tests and fixing existing tests, it seems to me we ought to make it as easy as possible to contribute.

**Comment by jscholes** [comment]
@zcorpan 

> It would be nice to automate tests earlier, why not in the Draft Test Phase?

Synthesizing the data needed to automate a test plan is potentially a lot of work.  We're not 100% clear on how much work, or exactly what the work will be just yet.  But it may involve:

1. recording expected output, for multiple screen readers;
2. recording a programmatic indication of how output should be parsed to determine the pass/fail status of assertions, plus presence of any unexpected behaviours;
3. programmatically matching parts of assertions to aspects of output (e.g. this substring represents the role); and
4. testing to ensure everything is working as expected.

Of course, not everything on this list will need to happen for every test plan.  For instance, many will reuse a common format for standardised assertions, e.g. relating to role and name.  But it will still represent a significant amount of effort, and it is suboptimal for that effort to be extended before the Community Group has reached consensus on a test plan during the draft phase:

* The best case scenario is that no changes are required, to either the test plan or its automation.
* In the worst case, the CG may decide to scrap a draft plan altogether, meaning that any automation-related input is wasted.

I'd guess that the most common case will fall between these two.  I.e. most test plans will result in a small amount of feedback and necessary changes during the draft phase, rather than 100% perfection or rejection.  But even there, those changes may also trigger necessary modifications to the automation-related assets.

**Comment by jscholes** [comment]
#### Some thoughts on the Glossary

* Assertion:
    * "Assertion Priority" is defined under the next heading, as a separate term.  But the priority is very much a mandatory part of an assertion, not a sibling of it.  I'd suggest either collapsing these two down into one, or having three: "Assertion Text", "Assertion Priority" and "Assertion", with the last of those just bringing together the first two under one umbrella.
    * The example assertion is worded as: "The screen reader conveys the role checkbox to the user.".  This is far more verbose than the format used within current tests, meaning that in reality it would be: "Role 'checkbox' is conveyed".  We also don't use a trailing period after assertions at present.  Of course, there could be a desire to expand the explicit indication of scope of assertions at some point, e.g. to separate screen-reader-related ones from assertions targeting other ATs?
* Assertion Priority: there are two dashes in "ARIA-AT".
* Test: this definition includes the task, commands and assertions.  It does not include any references to applicable modes, setup scripts and their descriptions, tester instructions, or references.  This may be out of scope for this page or section.

**Comment by jscholes** [comment]
#### Notes from August 19th CG Meeting

* We'd like a Group Operations page, that describes how the CG operates across the various workstreams, with list of contributors so people attending CG meeting can determine who's who.
* Add the word "plan" to phases ("Draft Test Plan Phase", "Candidate test Plan phase", "Recommended test Plan phase").
* What reasons could there be for sunsetting a test plan, vs results becoming outdated?  E.g. if an APG pattern itself is sunsetted; should we redirect users to an alternative?  The example given was Select-Only Combobox instead of Collapsible Listbox.
* Once people are happy with the new Working Mode, we need to decide what implications it has on other areas, e.g. will we change the columns on the GitHub project board to match these phases, or individual steps within them?  Also need to plan impact on app development.
* For Draft Test Phase, step 8:
    
    > (Community Group) Review and run the tests in the draft test plan and raise issues if the plan is incomplete, inaccurate, overly restrictive, excessively prescriptive, does not satisfy editorial conventions, contains scripting bugs, or otherwise does not meet community group standards for test plans.
    
    We would like more documentation around the meaning of the individual aspects reviewers should be looking out for, maybe including examples?  But we also don't want reviewers to just treat the review like a checklist; we welcome them following their gut and being opinionated when reading through test plan content.
* For Draft Test Phase, step 9:
    
    > (test developer) Resolve issues raised against draft test plan. If an issue is due to a problem with the ARIA specification, the APG, a browser, or an OS-level accessibility API, then:
    > * File an issue for the ARIA Working Group or APG Task Force as appropriate.
    > * ...
    
    Should we put this responsibility solely on the ARIA WG or APG Taskforce?  It may improve velocity if members of the ARIA-AT project are to file bugs directly (e.g. with browser bug trackers), as well.
* If an assertion fails due to an upstream dependency, we need to annotate that result with the reason, e.g. a URL to a Firefox bug, to avoid people assuming that an assistive technology is the blocker by default.  Those annotations could be removed automatically if automation reruns the test and determines that support is now good.  This is instead of just removing the assertion outright, which hides useful data.
* Possibility of automating test plan movement through different phases, e.g. if test plan has been in candidate phase for at least 120 days and has 0 open issues, automatically make it recommended.
* If an AT vendor opens issue during candidate phase to request more review time, should there be a constraint on that extension or the number that can be requested?  The feeling was that we should just use our best judgement on a case-by-case basis.
* Have a way of tracking latest browser and AT versions, preferably automatically via release feeds/APIs.  As soon as a new version of a browser and/or AT is known to be released, each recommended test plan run must be re-executed via automation to determine if the new version(s) is/are comparable with the old one.

CC @mcking65

**Comment by zcorpan** [comment]
@jscholes yes, good point that automating a test is more work for the test developer. However, running tests manually is also work. So I guess it's a trade-off, and maybe in some cases it would be less work overall to automate a test earlier.

**Comment by jscholes** [comment]
@zcorpan 

> running tests manually is also work

But that's work that can't be avoided, at least not entirely.  We will need some form of manual effort for every new test plan, in order to know what represents "good" output.  That may be manual testing with some automation built in (e.g. to gather speech), and/or manual review of human-generated test results to extract screen reader output.

**Comment by mcking65** [comment]
I have made revisions to address most feedback in this issue and promoted our draft to become the new [Working Mode](https://github.com/w3c/aria-at/wiki/Working-Mode).

Some of the feedback in this issue speaks to requirements we need to place on the aria-at app or to other more detailed documentation needs. I have included a link to this issue at the top of the working mode page in the "About this page" section so we can more easily follow up on those aspects of the feedback.

One dimension of feedback that I did not address with revisions to the document came from Simon. I think we should discuss it, though, because common understanding on the points he raises is important.

@zcorpan wrote:

> More high-level, I'm a bit concerned over the phases and the many steps. In my opinion, this was a problem with the previous working mode, too, but the new version doesn't fix it, instead adds to it.
> 
> If we want AT developers to contribute not only by reviewing tests, but also by writing new tests and fixing existing tests, it seems to me we ought to make it as easy as possible to contribute.

I've not envisioned AT developers contributing tests. This project is fundamentally different from WPT in that the tests for WPT are based on testing specified behavior. Anyone can write a test for WPT because it is a simple process to determine if the test itself is correct.

ARIA-AT tests, on the other hand, are used in lieu of a specification as the consensus vehicle. Developing an ARIA-AT test is more like developing a web specification. It is not easy. If the tests are to have meaning, they have to go through an appropriately rigorous process managed by a neutral party.

Given this, do the phases and steps seem appropriate?

When it comes to AT developer review of test during the candidate test review phase, I agree we need to do everything possible to make that process as approachable and efficient as possible. The goal of the working mode is to sketch what is needed without specifying precisely how it will be done.

I am closing this issue. We can continue any discussion that must happen here, but would prefer we open new issues for any discussion intended to generate further revisions to the working mode.
